src/services/shop/shop-router.ts

```diff
@@ -42,148 +43,423 @@ shopRouter.get(
+        responses: {
+            [StatusCode.SuccessOK]: {
+                description: "List of items and quantity",
+                schema: ShopOrderInfoSchema,
+            },
+            [StatusCode.ServerErrorInternal]: {
+                description: "Errors that should never happen",
+                schema: ShopInternalErrorSchema,
+            },
+        },
+    }),
+    async (req, res) => {
+        const { id: userId } = getAuthenticatedUser(req);
+
+        //get their order from order db
+        let userOrder = await Models.ShopOrder.findOne({ userId: userId });
+        if (!userOrder) {
+            const shopOrder: ShopOrder = {
+                items: [],
+                quantity: [],
+                userId: userId,
+            };
+
+            userOrder = await Models.ShopOrder.create(shopOrder);
+        }
+
+        if (!userOrder) {
+            return res.status(StatusCode.ServerErrorInternal).send(ShopInternalError);
+        }
+
+        return res.status(StatusCode.SuccessOK).send(userOrder);
     },
 );
 
-function getRand(index: number): string {
-    const hash = crypto.createHash("sha256").update(`${Config.JWT_SECRET}|${index}`).digest("hex");
-    return hash;
-}
+shopRouter.get(
+    "/cart/qr",
+    specification({
+        method: "get",
+        path: "/shop/cart/qr/",
+        tag: Tag.SHOP,
+        role: Role.USER,
+        summary: "Returns qr code of users cart",
+        responses: {
+            [StatusCode.SuccessOK]: {
+                description: "QR code",
+                schema: OrderQRCodesSchema,
+            },
+            [StatusCode.ClientErrorNotFound]: {
+                description: "Shop Item DNE",
+                schema: ShopItemNotFoundErrorSchema,
+            },
+            [StatusCode.ClientErrorBadRequest]: {
+                description: "Not enough quantity in shop",
+                schema: ShopInsufficientFundsErrorSchema,
+            },
+            [StatusCode.ClientErrorPaymentRequired]: {
+                description: "User doesn't have enough points to purchase",
+                schema: ShopInsufficientFundsErrorSchema,
+            },
+            [StatusCode.ServerErrorInternal]: {
+                description: "Errors that should never happen",
+                schema: ShopInternalErrorSchema,
+            },
+        },
+    }),
+    async (req, res) => {
+        const { id: userId } = getAuthenticatedUser(req);
+
+        // Fetch user order
+        const userOrder = await Models.ShopOrder.findOne({ userId });
+        if (!userOrder) {
+            return res.status(StatusCode.ServerErrorInternal).send(ShopInternalError);
+        }
+
+        const { items, quantity } = userOrder;
+
+        // Fetch all shop items in one query
+        const shopItems = await Models.ShopItem.find({ itemId: { $in: items } });
+        const itemMap = new Map(shopItems.map((item) => [item.itemId, item]));
+
+        // Validate item availability
```

Comment:
It's good that we do validation here, but /qr isn't the only thing that should do it. Make this functionality into a shared function so that it can be used for all endpoints to validate the order is still valid, and change it if not.